Quantization: preserving pre and post forward hooks

zetyquickly opened the pull request:
1. While do convert() preserve module's **pre and post forward** hooks
2. While do fusion preserve only module's **pre forward** hooks (because after fusion output no longer the same)

jerryzh168 commented:
Could you add some tests in quantization/tests_quantize.py?

zetyquickly commented:
@jerryzh168 
Yes, I thought about that. Will do in a while

zetyquickly commented:
@jerryzh168 
Added three test:
* On fusion. To check whether base models forward pre hooks move to parent which is fused module. (Base means first in fused list `Conv` in `[Conv, BN, ReLU]` for example) 
* Two test on preserving hooks while QAT and post-training statis is done

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -353,6 +353,12 @@ def swap_module(mod, mapping):
     if hasattr(mod, 'qconfig') and mod.qconfig is not None or type(mod) == DeQuantStub:
         if type(mod) in mapping:
             new_mod = mapping[type(mod)].from_float(mod)
+            # Preserve the module pre and post forward hooks except _observer_forward_hook
```
could you also write down when the hooks will be executed? that is: before/after quantized tensor

zetyquickly replied:
Updated

jerryzh168 commented on test/quantization/test_quantize.py:
```diff
@@ -447,6 +447,49 @@ def checkQuantized(model):
+        r"""Test post-training static quantization on preserving 
+        pre forward and post forward hooks of original model
+        """
+        for qengine in supported_qengines:
+            with override_quantized_engine(qengine):
```
you can use this decorator: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantized.py#L119

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -353,6 +353,13 @@ def swap_module(mod, mapping):
+            # Pre forward hooks will be called on quantized input
+            for pre_hook_fn in mod._forward_pre_hooks.values():
+                new_mod.register_forward_pre_hook(pre_hook_fn)
+            for hook_fn in mod._forward_hooks.values():
```
sorry, this one as well, I guess
before convert:
  this will be executed before the observer _forward_hook
and after convert:
  this will be executed after the tensor has been quantized?

jerryzh168 replied:
feels like we are not doing things consistently in this case, maybe the observer_forward_hook should run before these hooks as well

zetyquickly replied:
We have two options of calling observer_forward_hook:
1) When it's on Stub layer (that doesn't affect anything)
2) When it's on a non-stub layer (I mean Linear, Conv etc.) during static quantization

In the second case observer_forward_hook is registered for the examined layer first. And it will remain always first because of OrderedDict containing hooks

zetyquickly replied:
No, this is wrong. 
During preparation step observer_forward_hook will be the last hook in a row, so it will take into account any change that other hooks have already introduced to output tensor.
When we'll convert this module, the final output will be quantized properly. But as for output from hooks that maybe passed to other layers I am not sure.
What do you think?

raghuramank100 replied:
Thanks for this PR!. One thing to note is that for functional_modules hooks will not work (for both floating point and quantized modules).
https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/functional_modules.py

zetyquickly replied:
Yes, you're right.
Functional modules keep their behaviour as they've had previously

zetyquickly replied:
I think, the introduced code will preserve registered hooks on functional modules. But their forward is never called and they aren't meant to have hooks by construction

jerryzh168 commented on torch/quantization/fuse_modules.py:
```diff
@@ -124,6 +124,10 @@ def fuse_known_modules(mod_list):
         raise NotImplementedError("Cannot fuse modules: {}".format(types))
     new_mod = [None] * len(mod_list)
     new_mod[0] = fuser_method(*mod_list)
+    # Move pre forward hooks of the base module to fused module
```
what about post forward hook of the last module?

zetyquickly replied:
Yes, I don't see any restriction to implement that.
When I proposed that I thought that base layers such as Linear and Conv2d have more in common than ending layers activation and batchnorm

jerryzh168 replied:
I think it's still better to include that just for completeness

zetyquickly replied:
sure

jerryzh168 commented on test/quantization/test_quantize.py:
```diff
@@ -994,6 +1037,45 @@ def test_train_save_load_eval(self):
+
+                checkHooksIsPresent(model, True)
+                x = torch.rand(2, 5, dtype=torch.float)
+                _ = model(x)
```
nit: just `model(x)`

jerryzh168 commented on test/quantization/test_quantize.py:
```diff
@@ -994,6 +1037,45 @@ def test_train_save_load_eval(self):
+    def test_forward_hooks_preserved(self):
+        r"""Test QAT on preserving pre forward and post forward hooks of original model
+        """
+        for qengine in supported_qengines:
+            with override_quantized_engine(qengine):
```
same here to replace with decorators

Commit:
Quantization: preserving pre and post forward hooks

1. While do convert() preserve module's pre and post forward hooks
2. While do fusion preserve the first module's pre forward hooks and the last modules's post forward hooks

jerryzh168 reviewed (approved):
Looks good, thanks!

raghuramank100 commented on test/quantization/test_quantize.py:
```diff
@@ -1051,6 +1099,45 @@ def test_train_save_load_eval(self):
                 self.assertEqual(ref, out)
 
+    @override_qengines
+    def test_forward_hooks_preserved(self):
```
Can you also add a test for quantize_dynamic() API?

zetyquickly replied:
yes, sure. done

zetyquickly replied:
@raghuramank100 ping :)

raghuramank100 reviewed (changes requested):
Request additional tests, thanks for the PR!

jerryzh168 reviewed (approved):
lg

jerryzh168 commented:
got this error:     class TestPostTrainingStatic(QuantizationTestCase):
  File "/mnt/xarfuse/uid-30041/e9a1e41e-ns-4026533223/quantization/test_quantize.py", line 448, in TestPostTrainingStatic
    @override_qengines
NameError: name 'override_qengines' is not defined, could you import override_qengines from common_quantization.py?

zetyquickly commented:
> got this error: class TestPostTrainingStatic(QuantizationTestCase):
> File "/mnt/xarfuse/uid-30041/e9a1e41e-ns-4026533223/quantization/test_quantize.py", line 448, in TestPostTrainingStatic
> @override_qengines
> NameError: name 'override_qengines' is not defined, could you import override_qengines from common_quantization.py?

@jerryzh168 Fixed that. I've lost this during rebase somehow
